controller/engine_controller.go

```diff
@@ -1447,8 +1447,15 @@ func (ec *EngineController) startRebuilding(e *longhorn.Engine, replica, addr st
 		}
 		defer engineClientProxy.Close()
 
+		rep, err := ec.ds.GetReplica(replica)
+		if err != nil {
+			log.WithError(err).Errorf("Failed to get rebuilding replica %v", replica)
+			ec.eventRecorder.Eventf(e, v1.EventTypeWarning, EventReasonFailedRebuilding, "Failed to get rebuilding replica %v: %v", addr, err)
+			return
+		}
+
 		// If enabled, call and wait for SnapshotPurge to clean up system generated snapshot before rebuilding.
-		if autoCleanupSystemGeneratedSnapshot {
+		if autoCleanupSystemGeneratedSnapshot && rep.Spec.HealthyAt != "" {
```

Comment:
@shuo-wu For snapshot purge, should we handle this check on the engine side? so the manager doesn't need to do any validation for this but rather rely on the error returned from `SnapshotPurge`?

Reply:
Does this really resolve the issue? If Longhorn sends purge requests to the replica with empty `Spec.HealthyAt`, the cmd should fail and this engine controller will retry it. As a result, the overall rebuilding progress is just delayed a little bit.

Reply:
After discussing with @shuo-wu , this should be finally reconciled, so it should not make the volume pending in attaching state. 

@cchien816 Could you clarify this? Thanks.

Reply:
The reason why the volume gets attaching/detaching:
1. the eviction mechanism blindly removes the healthy replica on the node the test pod resides on while the test directly removes 2 healthy replicas on other nodes.
2. the test tries to r/w data from the volume. But at this moment the volume just contains a rebuilding replica, which is invalid for IO. Then the engine marks it as ERR after the IO failure
3. Longhorn somehow deleted this only ERR replica. As a result, the volume does not contain any replica and Longhorn wrongly marks the volume as attaching or detaching state.

Reply:
Using PR code base to run `test_disk_eviction_with_node_level_soft_anti_affinity_disabled`, test case pass and yes , volume not pending in attaching state

Reply:
As discussed with @shuo-wu , this seems like the day 1 issue instead of causing by the `purge-before-rebuild` introduction, but we don't know if this did not fail before, probably because of fast rebuilding before evicting the last replica?

Reply:
For referenceUsing PR code base to run some new fail cases in master [pipeline](https://ci.longhorn.io/job/public/job/master/job/sles/job/amd64/job/longhorn-tests-sles-amd64/199/#showFailuresLink)

- test_rebuild_with_restoration
- test_engine_crash_for_restore_volume
- test_engine_crash_for_dr_volume

Reply:
Then, they are issues with other causes which we need to identify if they are not flaky. Please create issues for them after the analysis. Thanks, @cchien816